packages/qwik/src/optimizer/src/plugins/bundle-graph.ts

```diff
@@ -0,0 +1,121 @@
+import { getSymbolHash } from 'packages/qwik/src/core/qrl/qrl-class';
+import type { QwikBundleGraph, QwikManifest } from '../types';
+
+/**
+ * A function that creates a modified version of the bundle graph. Used to inject routes and their
+ * dependencies into the bundle graph.
+ *
+ * @public
+ */
+export type BundleGraphModifier = (
+  graph: QwikBundleGraph,
+  manifest: QwikManifest
+) => QwikBundleGraph;
+
+const dynamicTag = '<dynamic>';
+/**
+ * This creates a compact array of dependencies for each bundle. It also contains the symbols. The
+ * format is:
+ *
+ * ```
+ *   [...(bundleName: string, ...directImports: index[], ...dynamicImports: [-1, ...index[]] | [])]
+ * ```
+ *
+ * (index is the position of the dependency in the bundleGraph array)
+ *
+ * This format allows any string to denote a set of dependencies, which is useful for symbols and
+ * SPA paths.
+ */
+export function convertManifestToBundleGraph(
+  manifest: QwikManifest,
+  bundleGraphModifiers?: Set<BundleGraphModifier>
+): QwikBundleGraph {
+  let bundleGraph: QwikBundleGraph = [];
+  const graph = manifest.bundles;
+  if (!graph) {
+    return [];
+  }
+  const names = Object.keys(graph).sort();
+  const map = new Map<string, { index: number; deps: Set<string> }>();
+  const clearTransitiveDeps = (parentDeps: Set<string>, seen: Set<string>, bundleName: string) => {
+    const bundle = graph[bundleName];
+    if (!bundle) {
+      // external dependency
+      return;
+    }
+    for (const dep of bundle.imports || []) {
+      if (parentDeps.has(dep)) {
+        parentDeps.delete(dep);
+      }
+      if (!seen.has(dep)) {
+        seen.add(dep);
+        clearTransitiveDeps(parentDeps, seen, dep);
+      }
+    }
+  };
+  for (const bundleName of names) {
+    const bundle = graph[bundleName];
+    const index = bundleGraph.length;
+    const deps = new Set(bundle.imports);
+    for (const depName of deps) {
+      if (!graph[depName]) {
+        // external dependency
+        continue;
+      }
+      clearTransitiveDeps(deps, new Set(), depName);
+    }
+    const internalDynamicImports = bundle.dynamicImports?.filter((d) => graph[d]) || [];
+    // If we have a lot of dynamic imports, we don't know which ones are needed, so we don't add any
+    // This can happen with registry bundles like for routing
+    if (internalDynamicImports.length > 0 && internalDynamicImports.length < 10) {
```

Comment:
We need to keep the `deps.hasSymbols` logic as explained in https://github.com/QwikDev/qwik/pull/7417#issuecomment-2750991973

> > Ok, in https://github.com/QwikDev/qwik/pull/7453 I just removed all of it and only disallow more than 10 dynamic deps because we're probably not going to guess right. That works fine too it seems.

> We need the hasSymbols guard, otherwise we will prefetch external dynamic deps again (e.g. shiki). 10 sounds like a reasonable safeguard, but I wonder if it's needed at all. The dynamic deps graph seems to stop on its own after a certain number of nodes. There is still a bit of over-prefetching on qwik-ui but I'm quite sure it's because some static deps that have dynamic deps aren't being tree-shaken somehow.

Reply:
but then we won't preload dynamic imports for library components :-/

Reply:
Hmmm are you sure? I will need to test this. Shouldn't be an issue if libraries use `preserveModules`, but better to support both ofc.

Reply:
I have a better solution now! You can adjust the graph in your vite config now, and I already use it to clear the dynamic imports from qwik-city.
I think preloading shiki or threejs or whatever is good if it is actually needed, so in case it's problematic the dev can optimize manually. See the bundle-graph.unit tests.

Reply:
I think preloading shiki or threejs or similarly "big" libs can be tricky. We can provide a way to preload those, but it should be optional imo. Otherwise that's also a breaking change.

Reply:
So I think we need to exclude external deps by default.

Reply:
How about a setting for maximum preload size?

Reply:
Might not be needed at all?

Reply:
yes indeed, we should only care about our own qrl imports, the rest are normally static imports, and when dynamic they are on purpose so we shouldn't interfere.